### PR TITLE
Rum Injection POC

### PR DESCRIPTION
**Pull request by ericmustin**
### Summary 

This PR adds an experimental `Datadog::Contrib::Rack::RumInjection` middleware to the `rack` instrumentation, that, when enabled and configured correctly, injects an [`HTML Comment`](https://www.w3schools.com/html/html_comments.asp) containing the current Datadog `trace-id` and the `trace-time` of the current trace, into a subset of html responses deemed suitable for injection, and updates their `Content-Length` HTTP Response Header. This is done so that the [`browser-sdk`](https://docs.datadoghq.com/real_user_monitoring/installation/?tab=us) can determine the initial document page load time for a request, and connect the the trace with a frontend RUM session. The HTML Comment is formatted:

```
<!-- DATADOG;trace-id=13285696538842353519;trace-time=1596643940000 -->
```

This middleware also contains functionality for an html meta tag containing this information, but this is disabled currently.

##### Middleware insertion

With Rails, the middleware is added to the rack middleware stack automatically as part of the rails patching code. To enable all that is required is enabling RUM Injection via the `Datadog.configure` block

```ruby
    Datadog.configure do |c|
      c.use :rack, rum_injection_enabled: true, rum_cached_pages: ['/api', '/blog/**/**']
      c.use :rails
    end
```
With just Rack, or with Rack and a compatible web frameworks(Sinatra, Grape, etc), the `Datadog::Contrib::Rack::RumInjection` middleware must be added at the appropriate location to the rack middleware stack manually. The intention is that it should be the _last_ middleware in the stack, or essentially the _last_ middleware called before any web framework routing middleware. This is because some middleware, such as `Rack::Deflater` will compress or gzip the html response, and decompression/re-compressing the response would create significant performance overhead in a user's application. To avoid this scenerio, ensure that the call to `use Datadog::Contrib::Rack::RumInjection` is done after any other application middleware. Here's an example below:

```ruby
# config.ru example
require 'ddtrace'

Datadog.configure do |c|
  c.use :rack, options
end

use Datadog::Contrib::Rack::TraceMiddleware
use Rack::Deflater
use Datadog::Contrib::Rack::RumInjection

app = proc do |env|
  [ 200, {'Content-Type' => 'text/plain'}, ['OK'] ]
end

run app
```

Any responses that have been compressed or gzipped are detected via html headers and skipped.

##### Caching / Determining the appropriate responses eligible for html comment injection

The other major hurdle is determining _which_ responses are eligible for injecting an HTML Comment.  The requirements here are largely determined via HTTP Response headers, with the following requirements:

  - must be a `Content-Type` of `html` or `xhtml`
  - cannot have a `Content-Encoding` of `gzip`, `compress`, or `deflate`
  - cannot be an attachment (`Content-Disposition` `attachment`)
  - cannot be a streaming response ( `chunked` `Transfer-Encoding`, `Content-Type`: `text/event-stream`, or using a RoR's controller we can reasonably assume generates a streaming response)
  - cannot be `cached` html.
    - otherwise this would mean that HTML with a specific trace-id could be re-used to fulfil multiple requests at either the browser or CDN level, an create an inaccurate associate between a Frontend RUM session and an unrelated backend generated trace-id. 
    - The logic for determining this is still part of ongoing work but largely can be defined as "Having relevant `Cache-Control`, `Expires`, and CDN Specific HTTP Response headers that indicate a response will not be cached.
    - Additionally, there is a `rack` configuration option `rum_cached_pages` to allow users to define an array of paths (including globbed paths) that they are certain will be cached and should not have a trace-id injected into the HTML.
    - A lot of this code is imprecise and there will likely still be scenerios where a piece of cached html contains a stale trace-id. The understanding here is that this is expected to some degree and that while we apply a best efforts methodology on the backend, the `brower-sdk` and apm intake pipeline will contain additional/more robust heuristics for determining whether to associate a trace-id with a RUM Session.

##### Alternative Approaches

A lot of this work is largely because the [`Server-Timing`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server-Timing) HTTP Response Header is not supported on Safari and older IE. It is the "right" way to do server timing frontend/backend association though, and so it's worth thinking about whether we want to include some stubs or functionality to "enable"  usage of `Server-Timing` header in the event that our above approach proves unreliable or that `Server-Timing` gains larger browser adoption in the future.

##### misc

This is mainly a POC and is something we'll look to deploy internally to test it's robustness and iterate. That being said any and all feedback here on the approach is very much appreciated, I think the "rack middleware" is really the only way to do this robustly across a lot of frameworks at once, but having to be specific about the middleware stack order does add some concerns around user experience, so if there's a cleaner way here, please let me know.

### CODE REVIEW

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        private
+
+        def configuration
+          Datadog.configuration[:rack]
+        end
+
+        def should_inject?(headers, env, status)
+          status == 200 &&
+            !env[RUM_INJECTION_FLAG] &&
+            no_cache?(headers, env) &&
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:103]: Do we have to check if the request is uncacheable here?
Given we are inserting a uniquely generated id into the response body, the request will automatically be uncacheable.

If the response was initially cachaeable, it is now uncacheable after our changes.
If the response was initially uncachaeable, it will stay that way.

I think that if the user enabled RUM injection, they expect it work as intended, meaning each response is now unique an uncacheable.

This gives even more importance to the `Server-Timing` header, as I believe it is possible to cache the body and request re-validation of only the headers.

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            (headers[CONTENT_TYPE_HEADER].include?(HTML_CONTENT) ||
+            headers[CONTENT_TYPE_HEADER].include?(XHTML_CONTENT))
+        end
+
+        def attachment?(headers)
+          (headers && headers.key?(CONTENT_DISPOSITION_HEADER) && !headers[CONTENT_DISPOSITION_HEADER].nil?) &&
+            headers[CONTENT_DISPOSITION_HEADER].include?(ATTACHMENT)
+        end
+
+        def streaming?(headers, env)
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:132]: If my suggestion for wrapping the response body object, and prepending only a single line to the body is accepted, we can actually support streaming, thus this method would become unnecessary.

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+          head_start = html.index(HEAD_TAG_OPEN)
+          html.index(TAG_CLOSE, head_start) + 1 if head_start
+        end
+
+        def meta_tag_template(trace_id, unix_expiry_time)
+          %(<meta name="dd-trace-id" content="#{trace_id}" /> <meta name="dd-trace-time" content="#{unix_expiry_time}" />)
+        end
+
+        def html_comment_template(trace_id, unix_expiry_time)
+          %(<!-- DATADOG;trace-id=#{trace_id};trace-time=#{unix_expiry_time} -->)
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:236]: Just a curiosity: why is the current time called `unix_expiry_time` here?

**Comment by marcotc** [comment]
One other thing I forgot: I think that if [`rack.hijack?` is enabled](https://www.rubydoc.info/github/rack/rack/file/SPEC#label-Hijacking), we shouldn't try to inject the RUM snippet at this moment.

Rack hijack means that the IO object will be provided from the `rack.hijack_io` environment variable, but that's an arbitrary IO implementation provided by the application. I don't think we want to inject content into that IO stream without a very good idea of what exactly the application is doing.

Until we are sure that we can identify which implementation has hijacked Rack for the current response, and that the implementation is relevant to us (maybe it's part of a popular library, like Rails or Sinatra), we should probably don't want to touch hijacked responses.

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+            current_trace_id = get_current_trace_id
+
+            # do not inject if no trace
+            return result unless current_trace_id
+
+            # we need to insert the trace_id and expiry meta tags
+            updated_html = generate_updated_html(response, headers, current_trace_id)
+
+            # update content length and return new response if we don't fail on rum injection
+            if updated_html
```
[the remark this answers is not in the document]
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb:73]: So, there's a few things here

- There are a number of ways that a CDN or a Browser can make a determination about whether a response should be cached. `Surrogate-Control`, `Cache-Control`, and `Expires` for example all need to be considered. Within those Headers there are a number of permutations that can be used to indicate "cache" or "do not cache" at both the client and cdn level. Additionally, for `Cache-Control` it can contain multiple comma separated values. You can see some level of detail these would include here: https://docs.fastly.com/en/guides/configuring-caching#applying-different-cache-rules-for-fastly-and-browsers . But this is CDN Specific and would vary considerably from CDN provider to provider. So, we can't just simply change the cache-control header to a default value when we inject.

- Additionally, the requirements of the POC are that we shouldn't be modifying the cache-control headers at all, instead what we're doing is attempting to _only_ inject on traces that we can safely say are _not_ Cached. Cached pages are not suitable for Rum injection. This will be included in our feature documentation.

- Even then, given things like VCL: https://docs.fastly.com/en/guides/guide-to-vcl, we can safely assume that we are not always going to be able to make the proper determination about whether or not a response is cached at either the browser or cdn level. We simply won't be able to anticipate every setup. Give that constraint, for this POC we are making an explicit decision to:

  - _attempt to only inject on non-cached pages_
  - _allow a configuration option so users can specify specific pages that are cached and shouldnt have trace-id injected_ `DD_TRACE_CACHED_PAGES`. 
  - attach timing information into the HTML Comment and allow the frontend browser-sdk to make a determination about whether the "trace-id" in the html, if it is present, is valid.  This will involve both logic within the browser sdk and logic within our own backend pipeline after intake. I have some thoughts on this that are probably not suitable for discussion here and we can sync up on internally, but tl;dr the goal here is to have something that works reasonably well as part of a POC, with some margin for error, and iterate.
  - Long term (post POC, before GA) the understanding is that we will also allow users an html template helper so that they can set the trace-id as a meta tag within their html templates directly. This will allow users the ability to only inject trace id into html they know is not cached, instead of us having to guess.

So, in summary, we can't change `Cache-Control` header to `no-store`.

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 1 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb:103]: the requirements are that we only inject when it is uncache-able, and do not inject if it is cacheable.

- The documentation states:
  > The RUM Injection middleware will insert an HTML comment into only those responses that are `Content-Type` `html` or `xhtml`, can be reasonably determined to be non-cached html responses at either browser or CDN level, are not streaming responses, and are not compressed or gzipped at time of injection.

  We should probably make that more prominent but the idea isn't to modify their caching strategy but to work around it.

- `Server-Timing` header is unfortunately not implemented across all modern browsers(safari, older ie) and is therefore unsuitable. This is a determination that was made for the POC and isn't something that can change in the language implementation. Again, I have some thoughts here that are probably better to sync on internally.

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 2 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb:132]: for the POC it was decided not to support streaming.

Additionally rails has some pretty clear documentation around not modifying response bodies in rack middleware when streaming and that they don't intended to support middlewares that do that.

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 3 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb:236]: just bad naming, can update

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+        def injectable_html?(headers)
+          (headers && headers.key?(CONTENT_TYPE_HEADER) && !headers[CONTENT_TYPE_HEADER].nil?) &&
+            (headers[CONTENT_TYPE_HEADER].include?(HTML_CONTENT) ||
+            headers[CONTENT_TYPE_HEADER].include?(XHTML_CONTENT))
+        end
+
+        def attachment?(headers)
+          (headers && headers.key?(CONTENT_DISPOSITION_HEADER) && !headers[CONTENT_DISPOSITION_HEADER].nil?) &&
+            headers[CONTENT_DISPOSITION_HEADER].include?(ATTACHMENT)
```
[the remark this answers is not in the document]
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb:129]: sure, makes sense

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        rescue Exception => e # rubocop:disable Lint/RescueException
+          Datadog.logger.warn("Error updating html for rum injection #{e.class}: #{e.message} #{e.backtrace.join("\n")}")
+          return nil
+        end
+
+        def concat_html_fragments(response)
+          # aggregate the html into a complete document
+          # should have a max amount we parse here after which we give up
+          html_doc = nil
+          response.each do |frag|
```
[the remark this answers is not in the document]
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: So, I thought of something that might be a little easier, which is that, since we will always be injecting the HTML Comment at the very start of the response, we can call `.each` , .insert, and then just break after doing so....since we aren't planning to support streaming afaik, this might be a little simpler than wrapping in an object and delegating methods (i'm a little hesitant to take this approach in general), and we avoid loading everything into memory like in the original logic.

What do you think?

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        def compressed?(headers)
+          headers && headers.key?(CONTENT_ENCODING_HEADER) &&
+            (headers[CONTENT_ENCODING_HEADER].include?(COMPRESS) ||
+              headers[CONTENT_ENCODING_HEADER].include?(GZIP) ||
+              headers[CONTENT_ENCODING_HEADER].include?(DEFLATE))
+        end
+
+        def injectable_html?(headers)
+          (headers && headers.key?(CONTENT_TYPE_HEADER) && !headers[CONTENT_TYPE_HEADER].nil?) &&
+            (headers[CONTENT_TYPE_HEADER].include?(HTML_CONTENT) ||
```
[the remark this answers is not in the document]
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb:132]: makes sense, updated

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 3 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: updated

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+
+        def initialize(app)
+          @app = app
+        end
+
+        def call(env)
+          result = @app.call(env)
+          return result unless configuration[:rum_injection_enabled] === true
+
+          status, headers, response = result
```
[the remark this answers is not in the document]
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: 👍  updated

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+          end
+
+          # last check expires
+          if headers.key?(EXPIRES_HEADER) && !headers[EXPIRES_HEADER].nil?
+            # Expires=0 means not cacced
+            return true if headers[EXPIRES_HEADER] == '0'
+          end
+        end
+
+        def get_current_trace_id
```
[the remark this answers is not in the document]
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: agreed, updated

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+            else
+              Datadog.logger.debug('Rum injection unsuccessful')
+              return result
+            end
+          end
+
+          # catchall if an earlier conditional is not met
+          result
+        rescue Exception => e # rubocop:disable Lint/RescueException
+          Datadog.logger.warn("error checking rum injectability #{e.class}: #{e.message} #{e.backtrace.join("\n")}")
```
[the remark this answers is not in the document]
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: ah good catch...I've wrapped our code in a begin/rescue block , updated the rescue to `StandardError` and in the event of the rescue blocck i return the original `result`

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+              return updated_response.finish
+            else
+              Datadog.logger.debug('Rum injection unsuccessful')
+              return result
+            end
+          end
+
+          # catchall if an earlier conditional is not met
+          result
+        rescue Exception => e # rubocop:disable Lint/RescueException
```
[the remark this answers is not in the document]
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: 👍  updated

**Comment by ericmustin** [comment]
@marcotc i tried to cover everything here, i tried a slightly different approach to improving the memory usage concern, and had some lengthy explanations around the caching stuff that may not be clear, so lmk what u think, or ping and we can sync on this as well

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcotc) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: > we can call .each , .insert, and then just break after doing so

I'm not sure I follow it here. We should avoid, unless necessary, to call `response.each` in our middleware, as this will eagerly start loading the response. I'm not sure if there's a way to avoid calling `#each` without some form of wrapping. If your suggestion is able to fulfill this requirement, feel free to go ahead and implement it.

For wrapping, I was thinking of something like this, which [follows Rack's spec for the `body`](https://www.rubydoc.info/github/rack/rack/file/SPEC#label-The+Body):
```ruby
class RumBody
  def initialize(first_line, original_body)
    @original_body = original_body
    @new_body = Enumerator.new { |y|
      y << first_line
      original_body.each { |e| y << e }
    }
  end

  def each(&block)
    @new_body.each(&block)
  end

  def close
    @original_body.close
  end
end
```

[the remark this answers is not in the document]
Comment (marcotc) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: I've re-read the updated code, I'll leave feedback on the new implementation instead. This last comment ^ is slightly outdated, regarding my take on your approach (which I can now clearly see in code).

My suggestion above is still valid, but feedback on the implementation will be left on separate comments on the relevant lines.

[the remark this answers is not in the document]
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: @marcotc i think its probably better to use a wrapping object...maybe using a Delegator though so that we don't have to worry about making sure all the other methods behave correctly?

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+              html_comment = html_comment_template(trace_id, unix_time)
+
+              # update content length and return new response if we don't fail on rum injection
+              if body.respond_to?(:each)
+                # inject html comment into first available fragment and then end early so we do not
+                # iterate over entire response
+                body.each do |fragment|
+                  fragment.insert(0, html_comment)
+                  break
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:80]: `break` wouldn't work with the semantics of `#each`, as it would forever consume the first element:
```ruby
File.write('/tmp/test', "first\nsecond")
iter = File.open('/tmp/test').each_line
iter.each { break }
iter.each { |l| puts l }
#> second
```

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+                end
+
+                body.close if body.respond_to?(:close)
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:83]: We should only call `#close` after iterating over all elements from `#each`. Here we are calling it without knowing if all elements have been iterated over (as we are breaking after a single element).

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+          # if we detect Server Side Event streaming controller, assume streaming
+          defined?(ActionController::Live) &&
+            env[ACTION_CONTROLLER_INSTANCE].class.included_modules.include?(ActionController::Live)
+        end
+
+        def no_cache?(headers, env)
+          # TODO: this is very complex, is there an easier way to determine cache behavior on cdn and browser
+          # TODO: glob performance may be worse than regex
+          return false unless configuration[:rum_cached_pages].none? do |page_glob|
+            File.fnmatch(page_glob, env['PATH_INFO']) if env['PATH_INFO']
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:165]: We should perform this check last, as this is likely a more expensive operation than the remaining HTTP response cache header checking we have to do.

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+
+          return false unless headers
+
+          # first check Surrogate-Controle, which fastly uses
+          # indicates a cdn cache if Surrogate-Control max-age>0,
+          # otherwise check cache-control, then expiry to determine if there is browser cache
+          return false if surrogate_cache?(headers)
+
+          # then check Cache-Control
+          if headers.key?(CACHE_CONTROL_HEADER) && !headers[CACHE_CONTROL_HEADER].nil?
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:176]:
```suggestion
          if cache_control_header = headers[CACHE_CONTROL_HEADER]
```

Let's also refer to `cache_control_header` instead of re-fetching it form the headers array again.

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+            return false if server_cache?(headers)
+
+            # then check max-age
+            if headers[CACHE_CONTROL_HEADER].include?(MAX_AGE)
+              # only not cached if max-age is 0
+              return headers[CACHE_CONTROL_HEADER].include?(MAX_AGE_ZERO)
+            end
+
+            # not cached if marked no-store, no-cache, or private
+            return (headers[CACHE_CONTROL_HEADER].include?(NO_CACHE) ||
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:189]: We should check these hard conditions first, as they are easier to help us short circuit out of this method.

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+          return false unless configuration[:rum_cached_pages].none? do |page_glob|
+            File.fnmatch(page_glob, env['PATH_INFO']) if env['PATH_INFO']
+          end
+
+          return false unless headers
+
+          # first check Surrogate-Controle, which fastly uses
+          # indicates a cdn cache if Surrogate-Control max-age>0,
+          # otherwise check cache-control, then expiry to determine if there is browser cache
+          return false if surrogate_cache?(headers)
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:173]: `Cache-Control` is a widespread concept, we should likely check it first, as it is more likely the the user application has modified these headers, over `Surrogate-Control`.

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+          (headers && headers.key?(CONTENT_DISPOSITION_HEADER) && !headers[CONTENT_DISPOSITION_HEADER].nil?) &&
+            !headers[CONTENT_DISPOSITION_HEADER].include?(INLINE)
+        end
+
+        def streaming?(headers, env)
+          # https://api.rubyonrails.org/classes/ActionController/Streaming.html
+          # rails recommends disabling middlewares that interact with response body
+          # when streaming via ActionController::Streaming
+          # TODO: if required to patch streaming, investigate patching further upstream, in the render action perhaps
+          return true if (headers && headers.key?(TRANSFER_ENCODING_HEADER) &&
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:148]: I see we are checking for the presence of `headers` many times (`if headers && ...`).
I think we should have a higher-level short circuit that skips pretty much all of this class if `headers` is not available.

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+      class RumInjection
+        include Datadog::Environment::Helpers
+
+        RUM_INJECTION_FLAG = 'datadog.rum_injection_flag'.freeze
+        INLINE = 'inline'.freeze
+        IDENTITY = 'identity'.freeze
+        HTML_CONTENT = 'text/html'.freeze
+        XHTML_CONTENT = 'application/xhtml+xml'.freeze
+        HEAD_TAG_OPEN = '<head'.freeze
+        TAG_CLOSE = '>'.freeze
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:25]: I noticed some of these are not needed anymore.

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+    module Rack
+      # RumInjection ensures that the Rack Response has rum information
+      # injected into it's html. The middleware modifies the response body
+      # of non-cached html so that it can be retrieved by the rum browser-sdk in
+      # the application frontend.
+      # rubocop:disable Metrics/ClassLength
+      class RumInjection
+        include Datadog::Environment::Helpers
+
+        RUM_INJECTION_FLAG = 'datadog.rum_injection_flag'.freeze
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:19]: How will RUM_INJECTION_FLAG be used?
I notice we only check for it, but never set it.
Is this an (undocumented) user settings?

Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb:18]: It gets set on the rack env 

here: https://github.com/DataDog/dd-trace-rb/pull/1139/files/68f0092be377639661137874974daa0594e1ba59#diff-e49d67d2fbda9897a00cada5fc03f70aR87

and we don't inject if it's set

check here: https://github.com/DataDog/dd-trace-rb/pull/1139/files/68f0092be377639661137874974daa0594e1ba59#diff-e49d67d2fbda9897a00cada5fc03f70aR115

The idea is to ensure the middleware is not called twice for any reason or the html is not inserted twice. if it is attempted to be inserted twice for whatever reason, we do not re-inject again.

Eventually, when we have a helper for manual insertion in a template, we can then set this env var to ensure we don't also inject again here.

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 19 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: 👍  removed

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 18 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: makes sense, tried to clean this up

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 17 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb:173]: the ordering of the checks is intentional since cdns take some headers with higher precedence than others. In this case Surrogate-Control is highest priority for fastly. I've updated the code comments here to be more explicitly about this

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 16 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: see other comment re: the headers have an intentional priority ordering. I've update the code comments to describe this more appropriately for future developers

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 15 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb:176]: 👍 makes sense. Rubocop yells at me when i do assignment in a conditional so can't be _as_ DRY as i'd like, but cleaned this up to use a variable

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 14 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: Makes sense, I just went ahead and moved it into it's own function that gets called only if `no_cache?` passes

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 13 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: updated this to use the wrapped obj approach

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (ericmustin) [reply lib/ddtrace/contrib/rack/rum_injection.rb]: i updated this to use the wrapped Body approach. Originally i had tried using `SimpleDelegator` but was hitting some issues on old versions of ruby, so went with the approach described above, tried to match the spec behavior.

**Comment by ericmustin** [comment]
@marcotc I tried to address all the feedback here. One thing I did notice is that when running some tests on a local rails application, `env['rack.hijack?']` was returning true by default. I'm not sure whether this due to running in development env, but I've disabled the early return for `env['rack.hijack?'] == true` for now.

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 15 above)
Comment (marcotc) [reply lib/ddtrace/contrib/rack/rum_injection.rb:171]: Rubocop won't yell if you put it inside parenthesis I believe (`if (cache_control_header = headers[CACHE_CONTROL_HEADER])`).

File: lib/ddtrace/contrib/rack/rum_injection.rb [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+        def initialize(app)
+          @app = app
+        end
+
+        def call(env)
+          result = @app.call(env)
+
+          begin
+            # TODO: env['rack.hijack?'] is true in rails tests
+            # dont check until we have more clarity on usage
```
Comment (marcotc) [inline lib/ddtrace/contrib/rack/rum_injection.rb:51]: Let's remove this for now, as `rack.hijack?` only tells us that Rack itself supports hijacking.
It doesn't seem like we can detect a Rack hijack from the variable in the spec.
We should still be safe overriding the body in a spec-compliant way, which we are doing.

File: lib/ddtrace/contrib/rack/rum_injection.rb (hunk 20 above)
Comment (marcotc) [reply lib/ddtrace/contrib/rack/rum_injection.rb:18]: 👍

File: docs/GettingStarted.md [hunk 22]
```diff
@@ -1169,6 +1170,8 @@ Where `options` is an optional `Hash` that accepts the following parameters:
 | `quantize.query.exclude` | Defines which values should be removed entirely. Excludes nothing by default. May be an Array of strings, or `:all` to remove the query string entirely. Option must be nested inside the `query` option. | `nil` |
 | `quantize.fragment` | Defines behavior for URL fragments. Removes fragments by default. May be `:show` to show URL fragments. Option must be nested inside the `quantize` option. | `nil` |
 | `request_queuing` | Track HTTP request time spent in the queue of the frontend server. See [HTTP request queuing](#http-request-queuing) for setup details. Set to `true` to enable. | `false` |
+| `rum_injection_enabled` | Connect frontend traces from the RUM (real user monitoring) [`browser-sdk`](https://docs.datadoghq.com/real_user_monitoring/installation/?tab=us) to backend traces by automatically injecting an HTML Comment containing the datadog trace-id. See [RUM Injection](#rum-injection) for setup details. Set to `true` to enable. Environment Variable: `DD_RUM_INJECTION`. *Experimental* | `false` |
```
Comment (marcotc) [inline docs/GettingStarted.md:1173]: Let's clarify that this only applies to html and xhtml pages that are non-cacheable, as per http response headers.

File: docs/GettingStarted.md [hunk 23]
```diff
@@ -1169,6 +1170,8 @@ Where `options` is an optional `Hash` that accepts the following parameters:
...
+| `rum_cached_pages` | Define which pages to exclude from automatically injecting an HTML Comment containing the Datadoog trace-id.  See [RUM Injection](#rum-injection) for setup details. Accepts an array of path values, including globbed paths e.g. `['/admin', 'api/**/update']`. Environment Variable: `DD_TRACE_CACHED_PAGES` , accepts an csv formatted string. *Experimental* | `[]` |
```
Comment (marcotc) [inline docs/GettingStarted.md:1174]: I was thinking that `rum_injection_disabled_paths` would be a more intuitive name, give the main option is called `rum_injection_enabled`. Also, the reason the user wants to disable RUM on these paths could be unrelated to caching, even though caching is likely the main reason.

What do you think?

**Comment by marcotc** [comment]
@ericmustin could you also take a look at the commit I pushed, c5bc521? I tried my best to not change any logic, but if there's anything there you don't like please let me know (or just override it 🙂).

**Comment by ericmustin** [comment]
@marcotc looked over your tweaks they all lgtm, didn't know about the `()` trick for rubocop 😅 . I  think the other feedback makes sense, i can push up those changes the other small details is that `DD_TRACE_CACHED_PAGES` is part of the POC spec across languages, so i can't modify that env var but can modify the setting to `rum_injection_disabled_paths`

File: docs/GettingStarted.md (hunk 23 above)
Comment (marcotc) [reply docs/GettingStarted.md:1174]: After talking with @ericmustin, we decided to keep these names as is, as they match what other Datadog tracer languages are implementing it.

File: docs/GettingStarted.md (hunk 22 above)
Comment (ericmustin) [reply docs/GettingStarted.md]: 👍 updated

**Comment by ericmustin** [comment]
@marcotc the tests had hung for some reason and were stuck as in-progress, and in the process of kicking them it would appear i have accidentally dimissed your review :( , if you wouldn't mind re-approving
